pandas/core/indexes/category.py

```diff
@@ -494,22 +493,27 @@ def _maybe_cast_indexer(self, key) -> int:
     def _get_indexer(
         self,
         target: Index,
-        method: Optional[str] = None,
-        limit: Optional[int] = None,
+        method: str | None = None,
+        limit: int | None = None,
         tolerance=None,
     ) -> np.ndarray:
+        # returned ndarray is np.intp
 
         if self.equals(target):
             return np.arange(len(self), dtype="intp")
 
         return self._get_indexer_non_unique(target._values)[0]
 
     @Appender(_index_shared_docs["get_indexer_non_unique"] % _index_doc_kwargs)
-    def get_indexer_non_unique(self, target):
+    def get_indexer_non_unique(self, target) -> tuple[np.ndarray, np.ndarray]:
+        # both returned ndarrays are np.intp
         target = ibase.ensure_index(target)
         return self._get_indexer_non_unique(target._values)
 
-    def _get_indexer_non_unique(self, values: ArrayLike):
+    def _get_indexer_non_unique(
+        self, values: ArrayLike
+    ) -> tuple[np.ndarray, np.ndarray]:
+        # both returned ndarrays are np.intp
```

Comment:
No need to add the return type to the function signature now, as the overloads will be used anyway, but I guess it's OK to keep

----

e.g.:
```python
from typing import overload, Literal, Optional, Union


@overload
def foo(bar: int) -> float: ...

@overload
def foo(bar: str) -> None: ...



def foo(bar: Union[str, int]): ...


bar: Union[int, str]
reveal_type(foo(bar))  # Revealed type is 'Union[builtins.float, None]'
```

Reply:
I think it is best to always include the types in the actual signature as well so that mypy checks the consistency of the function itself and reports an error with...

```
def foo(bar: Union[str, int]) -> float | None:
    return 42
```

Reply:
That's a good point, thanks!